scalene/scalene_asyncio.py

```diff
@@ -0,0 +1,205 @@
+import threading
+import gc
+
+from types import (
+    AsyncGeneratorType,
+    FrameType
+)
+from typing import (
+    List,
+    Tuple,
+    cast,
+)
+
+
+class ScaleneAsyncio:
+    """Provides a set of methods to collect idle task frames."""
+
+    should_trace = None
+    loops: List[Tuple[asyncio.AbstractEventLoop, int]] = []
+    current_task = None
+
+    @staticmethod
+    def current_task_exists(tident) -> bool:
+        """Given TIDENT, returns true if a current task exists.  Returns
+        true if no event loop is running on TIDENT."""
+        current = True
+        for loop, t in ScaleneAsyncio.loops:
+            if t == tident:
+                current = asyncio.current_task(loop)
+                break
+        return bool(current)
+
+    @staticmethod
+    def compute_suspended_frames_to_record(should_trace) -> \
+            List[Tuple[FrameType, int, FrameType]]:
+        """Collect all frames which belong to suspended tasks."""
+        # TODO this is an ugly way to access the function
+        ScaleneAsyncio.should_trace = should_trace
+        ScaleneAsyncio.loops = ScaleneAsyncio._get_event_loops()
+
+        return ScaleneAsyncio._get_frames_from_loops(ScaleneAsyncio.loops)
+
+    @staticmethod
+    def _get_event_loops() -> List[Tuple[asyncio.AbstractEventLoop, int]]:
+        """Returns each thread's event loop. If there are none, returns
+        the empty array."""
+        loops = []
+        for t in threading.enumerate():
+            frame = sys._current_frames().get(t.ident)
+            if frame:
+                loop = ScaleneAsyncio._walk_back_until_loop(frame)
+                # duplicates shouldn't be possible, but just in case...
+                if loop and loop not in loops:
+                    loops.append((loop, cast(int, t.ident)))
+        return loops
+
+    @staticmethod
+    def _walk_back_until_loop(frame) -> asyncio.AbstractEventLoop:
+        """Helper for get_event_loops.
+
+        Walks back the callstack until we are in a method named '_run_once'.
+        If this becomes true and the 'self' variable is an instance of
+        AbstractEventLoop, then we return that variable.
+
+        This works because _run_once is one of the main methods asyncio uses
+        to facilitate its event loop, and is always on the stack while the
+        loop runs."""
+        while frame:
+            if frame.f_code.co_name == '_run_once' and \
+               'self' in frame.f_locals:
+                loop = frame.f_locals['self']
+                if isinstance(loop, asyncio.AbstractEventLoop):
+                    return loop
+            else:
+                frame = frame.f_back
+        return None
+
+    @staticmethod
+    def _get_frames_from_loops(loops) -> \
+            List[Tuple[FrameType, int, FrameType]]:
+        """Given LOOPS, returns a flat list of frames corresponding to idle
+        tasks."""
+        return [
+            (frame, tident, None) for loop, tident in loops
```

Comment:
So each task frame gets returned with the `tident` of the thread running it?  Did you review how `tident` is used in the code to see if that still makes sense for task frames?